### DISCUSSION TITLE
Is there a way to disable\customize <a17-title-editor/> in 'layouts\form.blade.php'('Missing title' issue)?

### DISCUSSION THREAD

**Discussion by ammma** [area17/twill#1063] [Q&A]
From source code it looks like this component can not be customized or disabled but i'm facing many cases where this feature is not needed because it just doesn't make sense. Currents option i've found to deal with it is as follows: 
1. Set whatever field for `protected $titleColumnKey` and explain customer to not pay attention to information\errors appearing there. Not so easy considering it's always there on top of the form and showing 'Missing title' error while field is blank.
2. Make form to be shown in modal by enabling `'editInModal' => false` in controller's `protected $indexOptions`. Then Twill will use `create.blade.php` instead of 'form.blade.php' for edit so component `<a17-title-editor/>` won't be used. But here i'm facing numerous issues like: media field doesn't save, custom fields doesn't work, connected field doesn't work.

Examples: 
- entity with select and datetime fields. Date for a title doesn't make sense. And if you put related field(from select) it will show you value you don't need to see.
- entity `User` with fields `firstname` and `lastname`. Logically i would put them both but it is not possible.

Am i using wrong approach an other Twill tools needs to be used to handle this? Did anyone encountered same problems?

**Comment by pboivin** [comment]
Hi @ammma, you bring up some interesting points. Here's my personal take on it:

I think in general, this can be solved by "content modelling", for lack of a better term. You may not need a `title` field, but probably something equivalent: `name`, `identifier`, something along those lines. If you are generating slugs for a module, that's probably your concept of `title` right there. As you pointed out, you can use `$titleColumnKey` and `$titleFormKey` on your controllers to reconfigure this.

To your examples:

> entity with select and datetime fields. Date for a title doesn't make sense. And if you put related field(from select) it will show you value you don't need to see.

This is interesting. Can you share a bit more on this use case? Can this be remodeled into a repeater field attached directly to a parent element?


> entity User with fields firstname and lastname. Logically i would put them both but it is not possible.

Very good point. Here's a quick workaround to use a "computed field" as title :


**resources/views/admin/participants/create.blade.php**

```php
@formField('input', [
    'name' => 'firstname',
    'label' => 'Firstname',
    'translated' => true,
    'required' => true,
])

@formField('input', [
    'name' => 'lastname',
    'label' => 'Lastname',
    'translated' => true,
    'required' => true,
])
```


**app/Http/Controllers/Admin/ParticipantController.php**

```php
class ParticipantController extends ModuleController
{
    protected $moduleName = 'participants';

    protected $titleColumnKey = 'fullname';

    protected $titleFormKey = 'firstname';
}
```


**app/Repositories/ParticipantRepository.php**

```php
class ParticipantRepository extends ModuleRepository
{
    /* ... */

    public function afterSave($object, $fields)
    {
        $object->fullname = "{$object->firstname} {$object->lastname}";
        $object->save();

        parent::afterSave($object, $fields);
    }
}
```

I've omitted permalinks for simplicity. Hope this can help get you started!

**Reply by ammma** [reply]
Thanks for workaround examples.
In other words "computed field" is simply additional field in DB table which denormalizes data in order to make this feature working?

For my first example: imagine model `Event` with properties `date`, `type`. How would you solve this by "content modeling"?

**Comment by pboivin** [comment]
Hey @ammma, hope your project is doing well!

You're right, my example denormalizes data with the `fullname` column, which may not be ideal. I believe you could use an attribute accessor to the same effect:

```php
class Participant extends Model
{
    // ...

    public function getFullnameAttribute()
    {
        return "{$this->firstname} {$this->lastname}";
    }
}
```

I'm trying to provide simple examples to get you going, but perhaps I'm not being as helpful as I think I am. In any case, thanks for your feedback :)

**Reply by ammma** [reply]
hi @pboi20 , thanks for trying to help me, i appreciate this.

My main point here is to ensure that there is no "normal" way possible, hacky way i can find myself no problem.

Regarding your last proposal to add custom attribute to model - this is what i've tried to do first and it doesn't work. So for now i'm basically stuck with this.
